File: packages/sdk/src/economic/connectors/ap2.ts
/**
 * Google AP2 (Agent Payments Protocol) connector.
 *
 * Reads AP2 mandate constraints (spending caps, categories, expiry)
 * and enforces Veto policies ON TOP of mandate constraints.
 * Veto is always more restrictive, never less (D4).
 *
 * Timing mode: pre-call — mandate metadata available at setup.
 * Note: v1 does NOT verify mandate signatures. AP2 card-based
 * crypto path is still maturing; signature verification is Phase 2.
 *
 * @module economic/connectors/ap2
 */

import type { EconomicContext, ProtocolConnector, EconomicDenialDetails } from '../types.js';

/**
 * AP2 mandate data shape (from mandate negotiation response).
 */
export interface AP2MandateData {
  /** Mandate identifier */
  mandate_id?: string;
  /** Maximum spending amount for this mandate */
  spending_cap?: number;
  /** Amount already spent under this mandate */
  spent?: number;
  /** Mandate currency (ISO 4217) */
  currency?: string;
  /** Mandate expiry (ISO 8601) */
  expires_at?: string;
  /**
   * Allowed spending categories.
   *
   * Categories are advisory metadata in v1. Use Veto rules to enforce
   * tool-category constraints.
   */
  categories?: string[];
  /** Mandate signer (payer identity) */
  signer?: string;
  /** Per-action cost */
  cost?: number;
  /** Additional mandate metadata */
  [key: string]: unknown;
}

function extractMandateData(obj: Record<string, unknown>): AP2MandateData | null {
  const mandate = (obj.ap2_mandate ?? obj.mandate ?? obj) as Record<string, unknown>;

  const mandateId = mandate.mandate_id;
  if (typeof mandateId !== 'string' || !mandateId) return null;

  return {
    mandate_id: mandateId,
    spending_cap: typeof mandate.spending_cap === 'number' ? mandate.spending_cap : undefined,
    spent: typeof mandate.spent === 'number' ? mandate.spent : undefined,
    currency: typeof mandate.currency === 'string' ? mandate.currency : undefined,
    expires_at: typeof mandate.expires_at === 'string' ? mandate.expires_at : undefined,
    categories: Array.isArray(mandate.categories) ? mandate.categories.filter(c => typeof c === 'string') : undefined,
    signer: typeof mandate.signer === 'string' ? mandate.signer : undefined,
    cost: typeof mandate.cost === 'number' ? mandate.cost : undefined,
  };
}

function isMandateExpired(expiresAt: string): boolean {
  const expiry = new Date(expiresAt);
  // Fail-closed: treat unparseable dates as expired (payment boundary)
  if (Number.isNaN(expiry.getTime())) return true;
  return expiry.getTime() < Date.now();
}

function exceedsSpendingCap(cost: number, spendingCap?: number, spent?: number): boolean {
  if (spendingCap == null) return false;
  // If spent is unknown, assume 0 (fail-closed: enforce cap from scratch)
  return cost > (spendingCap - (spent ?? 0));
}

/**
 * Create an AP2 protocol connector.
 */
export function createAP2Connector(): ProtocolConnector {
  return {
    protocol: 'ap2',
    protocolVersion: '2026.03',

    extract(response: Response | Record<string, unknown>): EconomicContext | null {
      // Handle fetch Response objects
      if (response instanceof Response) {
        const mandateId = response.headers.get('x-ap2-mandate-id');
        const ap2Cost = response.headers.get('x-ap2-cost');
        const ap2Currency = response.headers.get('x-ap2-currency');
        const ap2Signer = response.headers.get('x-ap2-signer');
        const ap2SpendingCap = response.headers.get('x-ap2-spending-cap');
        const ap2Spent = response.headers.get('x-ap2-spent');
        const ap2Expires = response.headers.get('x-ap2-expires');

        if (!mandateId) return null;

        // Check mandate expiry
        if (ap2Expires && isMandateExpired(ap2Expires)) return null;

        const cost = ap2Cost ? Number(ap2Cost) : 0;
        if (!Number.isFinite(cost) || cost < 0) return null;

        // Enforce spending cap (D4: Veto is always more restrictive)
        const spendingCap = ap2SpendingCap ? Number(ap2SpendingCap) : undefined;
        const spent = ap2Spent ? Number(ap2Spent) : undefined;
        if (exceedsSpendingCap(cost, spendingCap, spent)) return null;

        return {
          cost,
          currency: (ap2Currency ?? 'USD').toUpperCase(),
          payer: ap2Signer ?? undefined,
          protocol: 'ap2',
          protocol_metadata: {
            mandate_id: mandateId,
            spending_cap: spendingCap,
            mandate_spent: spent,
            expires_at: ap2Expires ?? undefined,
          },
        };
      }

      // Handle plain objects (manual mode / pre-call)
      const mandate = extractMandateData(response);
      if (!mandate) return null;

      // Check mandate expiry
      if (mandate.expires_at && isMandateExpired(mandate.expires_at)) return null;

      // Enforce spending cap (D4: Veto is always more restrictive)
      const cost = mandate.cost ?? 0;
      if (exceedsSpendingCap(cost, mandate.spending_cap, mandate.spent)) return null;

      return {
        cost,
        currency: (mandate.currency ?? 'USD').toUpperCase(),
        payer: mandate.signer,
        protocol: 'ap2',
        protocol_metadata: {
          mandate_id: mandate.mandate_id,
          spending_cap: mandate.spending_cap,
          mandate_spent: mandate.spent,
          /**
           * Categories are advisory metadata in v1. Use Veto rules to
           * enforce tool-category constraints.
           */
          categories: mandate.categories,
          expires_at: mandate.expires_at,
        },
      };
    },
  };
}

/**
 * Build a connector_error denial for AP2 parse failures.
 */
export function buildAP2ConnectorError(error: string): EconomicDenialDetails {
  return {
    reason: 'connector_error',
    cost: 0,
    currency: 'USD',
    budget_scope: 'session',
    budget_limit: 0,
    budget_spent: 0,
    budget_remaining: 0,
    connector_name: 'ap2',
    raw_error: error,
  };
}

File: packages/sdk-python/veto/economic/connectors/ap2.py
"""Google AP2 connector."""

from __future__ import annotations

import math
from collections.abc import Mapping
from datetime import datetime, timezone
from typing import Any, Optional

from veto.economic.connectors._helpers import get_header
from veto.economic.types import EconomicContext, EconomicDenialDetails


def _number(value: Any) -> Optional[float]:
    if not isinstance(value, (int, float)) or isinstance(value, bool):
        return None
    numeric = float(value)
    return numeric if math.isfinite(numeric) else None


def _parse_number(value: Optional[str]) -> Optional[float]:
    if value is None:
        return None
    try:
        numeric = float(value)
    except ValueError:
        return None
    return numeric if math.isfinite(numeric) else None


def _is_expired(expires_at: str) -> bool:
    candidate = expires_at.strip()
    if candidate.endswith("Z"):
        candidate = f"{candidate[:-1]}+00:00"
    try:
        expiry = datetime.fromisoformat(candidate)
    except ValueError:
        return True
    if expiry.tzinfo is None:
        expiry = expiry.replace(tzinfo=timezone.utc)
    return expiry < datetime.now(timezone.utc)


def _exceeds_spending_cap(
    cost: float,
    spending_cap: Optional[float],
    spent: Optional[float],
) -> bool:
    if spending_cap is None:
        return False
    return cost > (spending_cap - (spent or 0))


def _extract_mandate_data(obj: Mapping[str, Any]) -> Optional[Mapping[str, Any]]:
    raw_mandate = obj.get("ap2_mandate", obj.get("mandate", obj))
    if not isinstance(raw_mandate, Mapping):
        return None
    mandate_id = raw_mandate.get("mandate_id")
    if not isinstance(mandate_id, str) or not mandate_id:
        return None
    return raw_mandate


class AP2Connector:
    protocol = "ap2"
    protocol_version = "2026.03"
    protocolVersion = protocol_version

    def extract(self, response: Any) -> Optional[EconomicContext]:
        mandate_id = get_header(response, "x-ap2-mandate-id")
        if mandate_id:
            expires_at = get_header(response, "x-ap2-expires")
            if expires_at and _is_expired(expires_at):
                return None

            cost = _parse_number(get_header(response, "x-ap2-cost")) or 0
            if cost < 0:
                return None

            spending_cap = _parse_number(get_header(response, "x-ap2-spending-cap"))
            spent = _parse_number(get_header(response, "x-ap2-spent"))
            if _exceeds_spending_cap(cost, spending_cap, spent):
                return None

            currency = get_header(response, "x-ap2-currency") or "USD"
            signer = get_header(response, "x-ap2-signer")
            return EconomicContext(
                cost=cost,
                currency=currency.upper(),
                payer=signer,
                protocol="ap2",
                protocol_metadata={
                    "mandate_id": mandate_id,
                    "spending_cap": spending_cap,
                    "mandate_spent": spent,
                    "expires_at": expires_at,
                },
            )

        if not isinstance(response, Mapping):
            return None

        mandate = _extract_mandate_data(response)
        if mandate is None:
            return None

        expires_at = mandate.get("expires_at")
        if isinstance(expires_at, str) and _is_expired(expires_at):
            return None

        cost = _number(mandate.get("cost")) or 0
        if cost < 0:
            return None

        spending_cap = _number(mandate.get("spending_cap"))
        spent = _number(mandate.get("spent"))
        if _exceeds_spending_cap(cost, spending_cap, spent):
            return None

        mandate_currency = mandate.get("currency")
        signer = mandate.get("signer")
        categories = mandate.get("categories")
        return EconomicContext(
            cost=cost,
            currency=mandate_currency.upper()
            if isinstance(mandate_currency, str)
            else "USD",
            payer=signer if isinstance(signer, str) else None,
            protocol="ap2",
            protocol_metadata={
                "mandate_id": mandate["mandate_id"],
                "spending_cap": spending_cap,
                "mandate_spent": spent,
                "categories": [item for item in categories if isinstance(item, str)]
                if isinstance(categories, list)
                else None,
                "expires_at": expires_at if isinstance(expires_at, str) else None,
            },
        )


def create_ap2_connector() -> AP2Connector:
    return AP2Connector()


def createAP2Connector() -> AP2Connector:
    return create_ap2_connector()


def build_ap2_connector_error(error: str) -> EconomicDenialDetails:
    return EconomicDenialDetails(
        reason="connector_error",
        cost=0,
        currency="USD",
        budget_scope="session",
        budget_limit=0,
        budget_spent=0,
        budget_remaining=0,
        connector_name="ap2",
        raw_error=error,
    )


def buildAP2ConnectorError(error: str) -> EconomicDenialDetails:
    return build_ap2_connector_error(error)
